### PR TITLE
Deserialization

### PR DESCRIPTION
**Pull request by patrycja-ziemkiewicz**
Implemented  deserialization: native types such as int, str, and float are first deserialized into Rust primitives and then converted into their corresponding Python types, while heap-allocated and collection types are deserialized directly into Python objects. 
Implemented RowsResult as an iterator over rows, each of which is mapped through the user-selected row representation via the row factory. 
Added full RowFactory support, with the default factory producing Python dicts and custom factories enabling arbitrary row shapes. 
Added ColumnIterator, which yields (column_name, column_value) pairs. When creating a RowsResult, the user may provide their own class extending RowFactory and implementing a build method. This method receives the ColumnIterator and can construct any custom row shape the user wants, using the iterator to access columns.
Added initial tests covering basic functionality.

### CODE REVIEW

File: src/value.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+// The trait does not provide a `Python` argument, meaning that Python objects which
+// would have to be constructed inside `deserialize()` or deeper in recursion
+// would require acquiring the GIL separately for every column/element/UDT/List
+// during nested decoding.
+//
+// Using our own `PyDeserializeValue` trait keeps all Python conversions inside
+// a single `Python::with_gil(...)` boundary, allowing complex values (lists,
+// sets, UDTs, nested collections) to be deserialized without repeated GIL
+// acquisition. This avoids potential slowdown of acquiring GIL multiples times.
+//
```
Comment (Lorak-mmk) [inline src/value.rs]: Note: We wouldn't neet to acquire and release GIL for each value.
We would need to acquire it once, and then in each value check that it is already acquired.

https://pyo3.rs/v0.27.1/performance.html#access-to-bound-implies-access-to-python-token

So the reasoning is fine, but the comment is a bit imprecise.

**Comment by wprzytula** [comment]
@patrycja-ziemkiewicz Please rebase on newest main, then reask for review.

File: src/value.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    ) -> Result<PyDeserializedValue<'py>, DeserializationError> {
+        let elem_typ = match typ {
+            ColumnType::Collection {
+                frozen: false,
+                typ: CollectionType::List(elem_typ),
+            }
+            | ColumnType::Collection {
+                frozen: false,
+                typ: CollectionType::Set(elem_typ),
+            } => elem_typ,
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/value.rs]:
```suggestion
        let elem_typ = match typ {
            ColumnType::Collection {
                frozen: _,
                typ: CollectionType::List(elem_typ),
            }
            | ColumnType::Collection {
                frozen: _,
                typ: CollectionType::Set(elem_typ),
            } => elem_typ,
```

File: src/value.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+impl<'frame, 'metadata, 'py, T> PyDeserializeValue<'frame, 'metadata, 'py> for Vec<T>
+where
+    T: PyDeserializeValue<'frame, 'metadata, 'py>,
+{
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue<'py>, DeserializationError> {
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/value.rs]: So this `Vec` serves just for static typing purposes, before we type-erase anyway into `PyAny`. Correct?

File: src/value.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+}
+impl<'py> IntoPyObject<'py> for PyDeserializedValue<'py> {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = Infallible;
+    fn into_pyobject(self, _: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(self.value)
+    }
+}
```
Comment (wprzytula) [inline src/value.rs]: :pick: missing empty line between `impl` blocks.

File: src/value.rs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<Self, DeserializationError> {
+        let cql = deser_cql_py_value(py, typ, v)?;
+
+        Ok(PyDeserializedValue { value: cql })
+    }
+}
```
Comment (wprzytula) [inline src/value.rs]: :bulb: A-ha! So this is the main point of interest to us. All implementation on collections that I saw above are just to be used in `deser_cql_py_value`. Neat and reasonable :+1:

File: src/value.rs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                Ok(PyInt::new(py, v).into_any())
+            }
+            NativeType::Text => {
+                let v = <&str as DeserializeValue<'frame, 'metadata>>::deserialize(typ, val)?;
+                Ok(PyString::new(py, v).into_any())
+            }
+            NativeType::Boolean => {
+                let v = bool::deserialize(typ, val)?;
+                Ok(PyBool::new(py, v).to_owned().into_any())
+            }
```
Comment (wprzytula) [inline src/value.rs]: :question: What does this `to_owned()` do here? Why is this needed specifically for boolean, while not needed for other primitive types?

File: src/value.rs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            _ => unimplemented!(),
+        },
+        ColumnType::Collection {
+            frozen: _frozen,
+            typ: col_typ,
+        } => match col_typ {
+            CollectionType::List(_type_name) => {
+                let list = Vec::<PyDeserializedValue>::deserialize_py(typ, val, py)?;
+                list.into_pyobject(py).map_err(DeserializationError::new)
+            }
```
Comment (wprzytula) [inline src/value.rs]: :thought_balloon: While this is still quite strange for me to see `Vec::deserialize_py` not deserialize to `Vec`, I really appreciate the tidyness of this solution. :100:

File: src/results.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+use stable_deref_trait::StableDeref;
+use std::fmt::Write;
+use std::ops::Deref;
+use yoke::{Yoke, Yokeable};
+
+// Originally written by @Lorak-mmk, moved by @patrycja-ziemkiewicz
+#[pyclass]
+pub(crate) struct RequestResult {
+    pub(crate) inner: scylla::response::query_result::QueryResult,
+}
```
Comment (wprzytula) [inline src/results.rs]: :wrench: Let's make sure this commit has `Co-authored-by` mention.

File: src/results.rs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+                match col {
+                    Some(c) => write!(result, "{}", c).unwrap(),
+                    None => write!(result, "null").unwrap(),
+                };
+                write!(result, "|").unwrap();
+            }
+            writeln!(result).unwrap();
+        }
+        Ok(PyString::new(py, &result))
+    }
```
Comment (wprzytula) [inline src/results.rs]: :open_mouth:  Wow, is this essentially a query result displayer? Do we want such thing? @Lorak-mmk

File: src/results.rs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+            row_col_iterator,
+            factory: f,
+        })
+    }
+}
+#[pyclass]
+pub struct RowsResult {
+    row_col_iterator: Py<RowColumnIterator>,
+    factory: Py<RowFactory>,
+}
```
Comment (wprzytula) [inline src/results.rs]: :pick: Missing empty line.

File: src/results.rs [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        slf
+    }
+}
+
+#[derive(Yokeable)]
+struct IteratorWrapper<'a> {
+    row_iterator: RawRowIterator<'a, 'a>,
+    column_iterator: ColumnIterator<'a, 'a>,
+}
+impl<'a> IteratorWrapper<'a> {
```
Comment (wprzytula) [inline src/results.rs]: :pick: Missing empty line.

File: src/results.rs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+// NOTE: In method attach_to_cart we cannot move out cart so We cannot use into_inner inside of it.
+// Fields in DeserializedMetadataAndRawRows are private We have to think
+// If it would be better to use method into_inner and construct new RawMetadata struct
+// or to implement some public method in rust driver that would give access to raw_rows.
+struct RawMetadata {
+    metadata: ResultMetadataHolder,
+    rows_count: usize,
+    raw_rows: Bytes,
+}
```
Comment (wprzytula) [inline src/results.rs]: :+1: `raw_rows` getter has been recently exposed for Node.js-RS Driver. I'm sure we can expose it for Python-RS as well.

File: src/results.rs [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+unsafe impl StableDeref for RawMetadataCart {}
+
+#[pyclass(name = "ColumnIterator")]
+pub struct RowColumnIterator {
+    yoked: Yoke<IteratorWrapper<'static>, RawMetadataCart>,
+}
```
Comment (wprzytula) [inline src/results.rs]: :question: Why is this called "RowColumnIterator"? What's a "RowColumn"? How do I parse this?

File: src/results.rs [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        slf
+    }
+}
+
+#[pyclass]
+pub struct Column {
+    #[pyo3(get)]
+    column_name: Py<PyString>,
+    #[pyo3(get)]
+    value: Py<PyAny>,
```
Comment (wprzytula) [inline src/results.rs]: :question: Can we make this typed? Don't we have a union type for all possible Python values deserializable from CQL?

File: python/tests/test_deserialization.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+async def test_simple_deserialization():
+    # 1. Connect
+    builder = SessionBuilder(["127.0.0.2"], 9042)
+    session = await builder.connect()
+
+    # 2. Create keyspace & table
+    await session.execute("""
+        CREATE KEYSPACE IF NOT EXISTS testks
+        WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
+    """)
```
Comment (wprzytula) [inline python/tests/test_deserialization.py]: :wrench: Please never use `SimpleStrategy` - just replace it with `NetworkTopologyStrategy`.

File: python/tests/test_deserialization.py [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    session = await builder.connect()
+
+    # 3. Create a table with complex CQL types
+    await session.execute("""
+                          CREATE TABLE IF NOT EXISTS testks.complex_table2 (
+                                                                              id int PRIMARY KEY,
+                                                                              name text,
+                                                                              scores list<int>,
+                          );
+                          """)
```
Comment (wprzytula) [inline python/tests/test_deserialization.py]: :wrench: Can we make this test use nested lists? Just to have this additional case tested.

**Review by wprzytula — COMMENTED** [review]
:100: Looks very promising - great work! :rocket:

File: src/value.rs [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+            let dict = PyDict::new(py);
+
+            for ((col_name, col_type), res) in iter {
+                let val = res.and_then(|v| {
+                    Option::<PyCqlValue>::deserialize_py(col_type, v.flatten(), py)
+                })?;
+
+                dict.set_item(col_name.to_string(), val).map_err(DeserializationError::new)?;
+            }
```
[the remark this answers is not in the document]
Comment (patrycja-ziemkiewicz) [reply src/value.rs]: Even though key/value conversion is infallible here, PyDict_SetItem can still fail at the Python runtime level.

File: src/value.rs [hunk 18]
```diff
@@ -0,0 +1,763 @@
...
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue, DeserializationError>;
+}
+
+pub struct List<T> {
+    phantom_data: std::marker::PhantomData<T>,
+}
```
Comment (Lorak-mmk) [inline src/value.rs]: Let's keep `pub` used only for items that are exposed to Python, and `pub(crate)` for items needed outside the module defining them. In other words, privacy of items should be as narrow as possible.
This is a general note, not only for those 2 items - I see this issue in other places, so please review your code and adjust.

File: src/value.rs
Comment (Lorak-mmk) [inline src/value.rs]: Commit: "Implement value deserialization logic"

In this commit `value.rs` is not included in the module tree (it is not defined in `lib.rs`), so its effectively a dead code. You could just fix it by including it here, but even then it will be effectively dead code. It seems to me that a better way to split this PR into commits would be the following:
- First perform the changes to `execute`, `RequestResult` etc. In other words, introduce the deserialization framework. This commit would include the traits, and other necessary things, but it would not include implementations of the trait, apart from let's say implementation for an int.
- Second commit adds all the other implementations, and expands type hints to include them.
- Third commit adds tests.

This would get rid of any dead code, and would make it easier to review the code: reader could immediately see how the traits you introduce are going to be used.

File: src/value.rs [hunk 19]
```diff
@@ -0,0 +1,763 @@
...
+                let v = Counter::deserialize(typ, val)?;
+                Ok(PyInt::new(py, v.0).into_any())
+            }
+            NativeType::Decimal => {
+                let dec = CqlDecimal::deserialize(typ, val)?;
+                let big: bigdecimal_04::BigDecimal = dec.into();
+                call_python_constructor(py, "decimal", "Decimal", (big.to_string(),), None)
+                    .map_err(DeserializationError::new)
+            }
+            NativeType::TinyInt => {
```
Comment (Lorak-mmk) [inline src/value.rs]: Could you check what python-driver and scyllapy deserialize cql Decimal to? Do they also use `decimal.Decimal`?

File: src/value.rs [hunk 20]
```diff
@@ -0,0 +1,763 @@
...
+                py_call_method(
+                    py,
+                    "datetime",
+                    "datetime",
+                    "fromtimestamp",
+                    (seconds,),
+                    Some(&kwargs),
+                )
+                .map_err(DeserializationError::new)
+            }
```
Comment (Lorak-mmk) [inline src/value.rs]: This also look unnecessarily complicated. PyO3 has support for chrono and time - can't we use one of those here?

File: src/value.rs [hunk 21]
```diff
@@ -0,0 +1,763 @@
...
+                let time: time_03::Time = t.try_into().map_err(DeserializationError::new)?;
+
+                let hour = time.hour();
+                let minute = time.minute();
+                let second = time.second();
+                let micro = time.nanosecond() / 1000;
+
+                call_python_constructor(py, "datetime", "time", (hour, minute, second, micro), None)
+                    .map_err(DeserializationError::new)
+            }
```
Comment (Lorak-mmk) [inline src/value.rs]: ditto

File: src/value.rs [hunk 22]
```diff
@@ -0,0 +1,763 @@
...
+            NativeType::Uuid => {
+                let v = uuid::Uuid::deserialize(typ, val)?;
+                call_python_constructor(py, "uuid", "UUID", (v.to_string(),), None)
+                    .map_err(DeserializationError::new)
+            }
+            NativeType::Timeuuid => {
+                let v = CqlTimeuuid::deserialize(typ, val)?;
+                call_python_constructor(py, "uuid", "UUID", (v.to_string(),), None)
+                    .map_err(DeserializationError::new)
+            }
```
Comment (Lorak-mmk) [inline src/value.rs]: Also please verify that python-driver and scyllapy use uuid for timeuuid.

File: src/value.rs (hunk 18 above)
Comment (Lorak-mmk) [inline src/value.rs]: What's the purpose of this type? Can't you implement your trait on PyList?

File: src/value.rs (hunk 6 above)
Comment (wprzytula) [reply src/value.rs]: This is OK IMO then.

File: src/value.rs [hunk 23]
```diff
@@ -0,0 +1,763 @@
...
+
+                let time: time_03::Time = t.try_into().map_err(DeserializationError::new)?;
+
+                let hour = time.hour();
+                let minute = time.minute();
+                let second = time.second();
+                let micro = time.nanosecond() / 1000;
+
+                call_python_constructor(py, "datetime", "time", (hour, minute, second, micro), None)
+                    .map_err(DeserializationError::new)
```
Comment (wprzytula) [inline src/value.rs]: ♻️ Could we maybe extract branches of this huge match to distinct functions? So that everything would get more modular.

Comment (wprzytula) [reply src/value.rs]: But it does make sense IMO to keep dispatcher (this function with the large `match`) separate from actual implementation. Jumpling between functions is only needed if you need to really read a particular implementation; what usually matters is ease of grasping the code top-down, which is easier if your functions are shorter and delegate things to intuitively extracted and named subroutines.

This is just my opinion, but I feel confident about it.

File: src/value.rs [hunk 24]
```diff
@@ -0,0 +1,763 @@
...
+                    .map_err(DeserializationError::new)?;
+
+                call_python_constructor(
+                    py,
+                    "dateutil.relativedelta",
+                    "relativedelta",
+                    (),
+                    Some(&kwargs),
+                )
+                .map_err(DeserializationError::new)
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/value.rs]: Perhaps it is. Here we handle errors differently (by returning custom DeserializationError) than we would in IntoPyObject impl AFAIU.

File: src/value.rs (hunk 22 above)
Comment (patrycja-ziemkiewicz) [reply src/value.rs]: They both use uuid.UUID.

File: python/scylla/_rust/results.pyi [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+    # - Vector
+    List["CqlValue"],
+
+    # CQL:
+    # - Set
+    Set["CqlValue"],
+
+    # CQL:
+    # - Tuple
+    Tuple["CqlValue", ...],
```
[the remark this answers is not in the document]
Comment (patrycja-ziemkiewicz) [reply python/scylla/_rust/results.pyi]: In legacy python driver it is deserialized to list.

File: src/deserialize/value.rs [hunk 26]
```diff
@@ -0,0 +1,721 @@
...
+                let Ok(obj) = map.into_pyobject(py);
+                Ok(obj)
+            }
+            // CQL Set → Python set
+            CollectionType::Set(_type_name) => {
+                let set = Set::<PyDeserializedValue>::deserialize_py(typ, Some(val), py)?;
+                let Ok(obj) = set.into_pyobject(py);
+                Ok(obj)
+            }
+            _ => unimplemented!(),
```
Comment (wprzytula) [inline src/deserialize/value.rs]: :wrench: Let's provide more context to this `unimplemented()` message.

File: src/deserialize/value.rs [hunk 27]
```diff
@@ -0,0 +1,721 @@
...
+                        Some(v) => PyDeserializedValue::deserialize_py(typ, Some(v), py),
+                        None => Ok(PyDeserializedValue::new(py_none(py))),
+                    });
+                PyValueOrError::new(result)
+            });
+
+            let tuple = PyTuple::new(py, t)?;
+            Ok(tuple.into_any())
+        }
+        _ => unimplemented!(),
```
Comment (wprzytula) [inline src/deserialize/value.rs]: Ditto: more context needed in `unimplemented!()`.

File: src/deserialize/conversion.rs [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+            add_base10(&mut output, &temp);
+        }
+
+        // base *= 256 for the next byte
+        mul_scalar_base10(&mut base, 256);
+    }
+
+    output.reverse();
+    output
+}
```
Comment (wprzytula) [inline src/deserialize/conversion.rs]: :warning: This allocates one `Vec` per each non-zero digit... This is horrible!

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: src/deserialize/conversion.rs (hunk 28 above)
Comment (patrycja-ziemkiewicz) [reply src/deserialize/conversion.rs]: I removed the per-digit Vec allocation, but as mentioned earlier this implementation is not optimal and can be further improved.

File: python/scylla/_rust/results.pyi [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+    # CQL:
+    # - Time
+    time,
+    # CQL:
+    # - Duration
+    relativedelta,
+    # CQL:
+    # - Empty
+    # - null
+    None,
```
Comment (Lorak-mmk) [inline python/scylla/_rust/results.pyi]: If this is only used for deserialization, then Empty can be dropped from the comment I think.
If this is used also for serialization, then we need a separate type for Empty.

File: python/tests/test_deserialization.py [hunk 30]
```diff
@@ -0,0 +1,1060 @@
...
+
+    for inner in nested1:  # pyright: ignore[reportUnknownVariableType]
+        assert isinstance(inner, list)
+
+    for inner in nested2:  # pyright: ignore[reportUnknownVariableType]
+        assert isinstance(inner, list)
+
+    # Value checks
+    assert nested1 == v1
+    assert nested2 == v2
```
Comment (Lorak-mmk) [inline python/tests/test_deserialization.py]: Please add a case with empty inner list.

File: python/tests/test_deserialization.py [hunk 31]
```diff
@@ -0,0 +1,1060 @@
...
+        (5, Decimal("-7890.00123")),
+        (6, Decimal("123456789012345678901234567890.123456789")),
+        (7, Decimal("-999999999999999999999999999999999999.9999")),
+        (8, Decimal("0.00000000000000000001")),
+        (9, Decimal("42.000000")),
+        (10, Decimal("617283694")),
+        (11, Decimal("617283694.4324")),
+    ],
+)
+async def test_decimal_deserialization(
```
Comment (Lorak-mmk) [inline python/tests/test_deserialization.py]: You wrote this test using parametrize, but in all other tests you duplicated all test lines for all cases.

Please remove the duplication in other tests. We don't want to have to duplicate whole test of e.g. int just to add a test case.
Wheter you use `parametrize` or just an array defined inside test is up to you, I don't have enough knowledge about Python testing conventions to decide that.

File: python/tests/test_deserialization.py [hunk 32]
```diff
@@ -0,0 +1,1060 @@
...
+    await session.execute(f"INSERT INTO {table} (id, ts) VALUES (2, {ts2});")
+
+    rows = list((await session.execute(f"SELECT * FROM {table}")).iter_rows())
+
+    r1 = next(r for r in rows if r["id"] == 1)
+    r2 = next(r for r in rows if r["id"] == 2)
+
+    assert r1["ts"] == dt1
+    assert r2["ts"] == dt2
+
```
Comment (Lorak-mmk) [inline python/tests/test_deserialization.py]: Again, please introduce regression test for edge cases. This comment applies also to other date/time related types where you used fallible conversion in deserialization.

File: src/deserialize/results.rs [hunk 33]
```diff
@@ -159,29 +151,31 @@ unsafe impl StableDeref for DeserializedMetadataAndRawRowsCart {}
...
-        let cart = DeserializedMetadataAndRawRowsCart(Box::new(data));
-
-        let yoked = Yoke::attach_to_cart(cart, |cart| {
-            let frame_slice = FrameSlice::new(cart.raw_rows());
+    fn new(query_result: Arc<QueryResult>) -> PyResult<Self> {
+        let cart = QueryResultCart(query_result);
+
+        let yoked = Yoke::try_attach_to_cart(cart, |cart| -> PyResult<_> {
+            let Some(raw_rows_with_metadata) = cart.deserialized_metadata_and_rows() else {
+                return Err(PyRuntimeError::new_err("Result does not have rows"));
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/deserialize/results.rs]: IMO iterating over non-rows result should be possible - such iteration should return None immediately. This is because we're in Python, where (I believe) we care mostly about ease of use, less about rigidity.

File: python/scylla/_rust/results.pyi (hunk 25 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply python/scylla/_rust/results.pyi]: But in Python there's no distinction tuple vs list in the aspect of heterogeneity. The main (maybe the only one) difference is in mutability, and I see Vector as a fixed length immutable type, which matches Python tuple.

File: src/deserialize/value.rs [hunk 34]
```diff
@@ -91,13 +516,190 @@ fn deser_cql_py_value<'py, 'metadata, 'frame>(
...
+                Ok(PyBool::new(py, v).to_owned().into_any())
+            }
+            // CQL Date → Python datetime.date
+            NativeType::Date => {
+                let date: chrono_04::NaiveDate = CqlDate::deserialize(typ, Some(val))?
+                    .try_into()
+                    .map_err(DeserializationError::new)?;
+
+                Ok(date.into_pyobject(py)?.into_any())
+            }
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/deserialize/value.rs]: +1 for opening the issue.

File: src/deserialize/conversion.rs [hunk 35]
```diff
@@ -0,0 +1,61 @@
...
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        let bytes = self.val.as_signed_bytes_be_slice();
+        unsafe {
+            let val = ffi::_PyLong_FromByteArray(bytes.as_ptr(), bytes.len(), 0, 1);
+
+            Ok(Bound::from_owned_ptr(py, val).downcast_into()?)
+        }
+    }
+}
```
Comment (wprzytula) [inline src/deserialize/conversion.rs]: :question: Where and how did you find this `ffi::_PyLong_FromByteArray`? I had trouble when perusing PyO3 docs.

File: src/value.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/value.rs]: It looks very good now.

File: src/value.rs (hunk 23 above)
Comment (wprzytula) [reply src/value.rs]: Now the branches are quite concise, so it's no longer relevant.

File: src/deserialize/value.rs [hunk 36]
```diff
@@ -0,0 +1,763 @@
...
+                kwargs
+                    .set_item("months", d.months)
+                    .map_err(DeserializationError::new)?;
+                kwargs
+                    .set_item("days", d.days)
+                    .map_err(DeserializationError::new)?;
+                kwargs
+                    .set_item("microseconds", d.nanoseconds / 1000)
+                    .map_err(DeserializationError::new)?;
+
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/deserialize/value.rs]: @Lorak-mmk WDYT?

File: src/deserialize/results.rs (hunk 33 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/deserialize/results.rs]: Please check what old Python driver does.

File: python/scylla/_rust/results.pyi (hunk 29 above)
Comment (Lorak-mmk) [reply python/scylla/_rust/results.pyi]: The database rejects empty values send by the driver?

Comment (wprzytula) [reply python/scylla/_rust/results.pyi]: OK, this is indeed supported for `CqlValue::Empty`. So `CqlValue` is the only way to insert empty CQL values.

Comment (wprzytula) [reply python/scylla/_rust/results.pyi]: We should probably implement `SerializeValue` for `MaybeEmptiable` then.
